halo2/src/recursion/endoscale.rs

```diff
@@ -0,0 +1,102 @@
+//! Gadget for endoscaling.
+
+use ff::PrimeFieldBits;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter},
+    plonk::Error,
+};
+use pasta_curves::arithmetic::CurveAffine;
+
+mod chip;
+pub mod primitive;
+
+/// Instructions to map bitstrings to and from endoscalars.
+pub trait EndoscaleInstructions<C: CurveAffine>
+where
+    C::Base: PrimeFieldBits,
+{
+    /// A bitstring to be used in endoscaling.
+    type Bitstring: std::fmt::Debug + Clone;
+    /// The maximum number of bits that can be represented by [`Self::Bitstring`].
+    const MAX_BITSTRING_LENGTH: usize;
+
+    /// Witnesses a slice of bools as a vector of [`Self::Bitstring`]s.
+    fn witness_bitstring(bits: &[bool]) -> Vec<Self::Bitstring>;
+
+    /// Computes commitment (Alg 1) to a variable-length bitstring using the endoscaling
+    /// algorithm. Uses a fixed base.
+    ///
+    /// The bitstring is decomposed into pairs of bits using a running sum outside of
+    /// this gadget.
```

Comment:
```suggestion
    /// The bitstring is decomposed into pairs of bits using a running sum outside of
    /// this instruction.
```
But I'd also argue that this is a boundary violation: the instruction user should not need to care how this is implemented, and a chip implementation should be able to decompose with running sum inside this instruction impl if it's better for their use case. If instead every avenue for constructing a `Bitstring` requires that the resulting type is (eventually) constrained, then this becomes an implementation detail.